### ISSUE TITLE
"Failed to execute 'postMessage' on 'DOMWindow'" with Turbo

### ISSUE THREAD

**Issue by sedubois** [stripe/react-stripe-js#188]
NB: this issue is not directly related to React Stripe, but I'm opening it here [as suggested by @sashko-stripe](https://github.com/stripe/react-stripe-elements/issues/98#issuecomment-808459130).

### Summary

After adding the Stripe JS SDK to a [Turbo](https://turbo.hotwire.dev/)-enabled app (previously named Turbolinks), this Javascript error appears when navigating within the app:
```
Failed to execute ‘postMessage’ on ‘DOMWindow’: The target origin provided (‘https://js.stripe.com’) does not match the recipient window’s origin (‘http://localhost:3000’).
```

### Other information

To reproduce:
- install Rails on the machine (tested with Rails 6.1.3.2 on macOS)
- `rails new --skip-javascript stripe-test-app && cd stripe-test-app`
- `echo 'gem "turbo-rails"' >> Gemfile` (tested with `turbo-rails` 0.5.9)
- `./bin/bundle install`
- `./bin/rails turbo:install`
- add `<script src="https://js.stripe.com/v3"></script>` to `app/views/layouts/application.html.erb`'s head
- `./bin/rails g scaffold Post title:string && ./bin/rails db:migrate` (to generate views with some links which can be navigated)
- `./bin/rails server`
- visit http://localhost:3000/posts (tested with Firefox 88)
- open the browser's javascript console
- click the link 'New Post'
- the error message should appear in the console after a couple of seconds

<img width="978" alt="Screenshot 2021-05-21 at 15 58 29" src="https://user-images.githubusercontent.com/4217871/119149063-700bc600-ba4d-11eb-8b9d-3ca635dc24f7.png">

https://user-images.githubusercontent.com/4217871/119150301-9aaa4e80-ba4e-11eb-8eb9-b06ec47d23c4.mov

**Comment by sedubois** [comment]
@sashko-stripe any suggestion?

**Comment by alec-c4** [comment]
Hey! Any news on this issue?

**Comment by sedubois** [comment]
@alec-c4 I'm not aware of any solution to this issue.

**Comment by alec-c4** [comment]
@sedubois I've found a temporary solution - I've disabled turbo-rails on pages where I use stripe.js. The problem is because of their conflict with each other.

**Comment by sedubois** [comment]
@alec-c4 but Stripe’s documentation states that the Stripe SDK should be loaded on every page of the website (e.g. to help with fraud prevention).

**Comment by alec-c4** [comment]
@sedubois do you admin pages too? I think it is required only for pages with a payment form

**Comment by sedubois** [comment]
> @sedubois do you admin pages too? I think it is required only for pages with a payment form

@alec-c4 see https://github.com/hotwired/turbo/issues/270#issue-898099374:

> The threads over at turbolinks/turbolinks#321 and hotwired/turbo#198 suggest to force a full reload for pages which include the Stripe JS SDK in order to avoid the error Failed to execute 'postMessage' on 'DOMWindow'. However Stripe's [documentation](https://stripe.com/docs/js/including) states:
> 
>     To best leverage Stripe’s advanced fraud functionality, include this script on every page, not just the checkout page. This allows Stripe to detect suspicious behavior that may be indicative of fraud as customers browse your website.
> 
> So the SDK needs to be included on every page, but forcing a page reload on every page is hardly practical. Are there any alternatives besides having this error appear after every single page change?

**Comment by bufordtaylor** [comment]
Is there a solution for this yet?

**Comment by sedubois** [comment]
@bufordtaylor see https://github.com/hotwired/turbo/issues/270#issuecomment-1068130327 for a workaround.

**Comment by codemonstur** [comment]
I have this issue during development in my app. The app is a plain vanilla JS website that uses some jquery and navigo for routing. I've written some test code that can be used to trigger the bug on my machine. See here: https://github.com/codemonstur/stripe-bug

Triggering it requires some work. First load the page in the browser, make sure you have a bunch of extensions enabled. Click back and forth on the page loading buttons. After a couple of clicks the error will show up.

It is reliable enough to work on my machine but I have noticed that if I turn all the extensions off the error no longer appears. At the same time there is no one extension that is the offending party. Instead the bug gets less and less triggerable the fewer extensions I have enabled. I've been able to get the bug using just adblock + webarchives, but also with cookie deleter + mercury reader + javascript switcher. There might be a timing / load factor involved with triggering this problem.

Separately I found a stackoverflow post about the exact same issue but for YouTube. The comments there mention that the error can happen when a script from an HTTPS page tries to inject an iframe without a scheme in the URL on a page that uses HTTP. This is the case during development for me because the Stripe JS is on HTTPS but my local development app is on HTTP.

https://stackoverflow.com/questions/27573017/failed-to-execute-postmessage-on-domwindow-https-www-youtube-com-http
